### PR TITLE
Applicative ignorance

### LINKED ISSUE #16: `++` is too noisy/verbose
```swift
one ++ two ++ three
```

is the concatenation of three parsers but you as much of the concatenation as the parsers.

```swift
one + two + three
```

is a bit better, and `+` already basically means concatenation for `SequenceType`.

We could also add overloads for arrays/various arities of tuples:

```swift
(one, two, three) --> { … }
```

would be an overload of `-->` applying to 3-tuples of parsers. This would, sadly, need a lot of overloads, but it would also be kind of nice because it would avoid the problem that `++` has (as would other infix operators) where you get a right-associative list of pairs instead of an n-tuple. That is,

```swift
a ++ b ++ c ++ d ++ e
```

results in a parse tree that looks like

```swift
(a, (b, (c, (d, e)))
```

which is really obnoxious to unpack: `d` is `$0.1.1.1.0` instead of, say, `$3`.

Since we probably don’t want to have overrides for every combination of operator, side, and tuple-arity, we could instead do something similar to #15 and use an operator:

```swift
%(a, b, c, d, e) --> { … }
```

### LINKED ISSUE #25: Ignore literals by default
`%"x"` produces `"x"` by default, which is actually kind of redundant. I tend to find myself doing `ignore("x")` most of the time instead. You can always recover the original behaviour, too:

```swift
ignore("x") --> const("x")
```

We could reintroduce `literal` as a shorthand for the non-ignored version.

### LINKED ISSUE #36: Infix repetition binds looser than concatenation
I.e. `x * 3 ++ y` won’t compile because it’s parsed as `x * (3 ++ y)` instead of `(x * 3) ++ y`.

### LINKED ISSUE #90: Use *> to concatenate parsers, dropping the left hand side’s trees
A la `Applicative`’s [`*>`](http://hackage.haskell.org/package/base-4.8.0.0/docs/Prelude.html#v:-42--62-) or `Monad`’s [`>>`](http://hackage.haskell.org/package/base-4.8.0.0/docs/Prelude.html#v:-62--62-). I think `*>` communicates more clearly for our purposes.

If there is a similar operator for `Alternative` then we can use it and this to get rid of `ignore` altogether.

### PR DESCRIPTION
**Pull request by robrix**
Your comments would be appreciated!

This replaces the various overrides of the `++` and `|` operators for concatenation and alternation:

- `<*>` concatenates.
- `<*` concatenates, dropping the output of the right hand side.
- `*>` concatenates, dropping the output of the left hand side.
- `<||>` alternates.
- `<||` alternates, dropping the output of the right hand side.
- `||>` alternates, dropping the output of the left hand side.

The overloads of `++` and `|` which ignored input on both sides are no longer required since this is no longer driven by the types, and instead by the selection of operator. In practice I’m finding it quite pleasant to use.

~~I’m not thrilled with the selection of `<||>` for alternation; by all rights this ought to be `<|>`, but then there’s ambiguity between the `|>` and `<|` alternation operators and the application operators in Prelude. However, since it is, in practice, quite uncommon to use `<||` and `||>`, we could change `<||>` to `<|>` and either a) leave them as `<||` and `||>` or b) remove them altogether (since their behaviour can be recovered fairly easily with e.g. `map`).~~

**Edit:** This now uses `<|>` for alternation and omits dropping output on either side. It can be recovered, if desired, with `map`, `<^>`, or `-->`.

- Fixes #16.
- Closes #25.
- Fixes #36.
- Fixes #90.

### CODE REVIEW

**Comment by robrix** [comment]
I’ve decided to restore `<|>`, and drop `||>` and `<||`.

**Comment by robrix** [comment]
Alright, ready for review :sparkles:

**Comment by robrix** [comment]
https://twitter.com/jspahrsummers/status/599374650259345408

**Comment by sharplet** [comment]
This looks pretty cool! I really like `<*` and `*>` (first time I've encountered them), and having a custom `<|>` with a lower precedence works nicely with the other operators.

Something did trip me up though — take this example in Colours.playground:

```swift
let hex2 = hex <*> hex |> map { $0 + $1 }
```

I thought I'd see if it could also be used like this:

```swift
let hex2 = { $0 + $1 } <^> hex <*> hex
```

Needs to be curried though:

```swift
let hex2 = curry({ $0 + $1 }) <^> hex <*> hex
```

But the type of this expression ends up being something like `Parser<String, (String -> String, String)>.Function`.

To make that expression work, the definition of `<*>` would have to be something like this:

```swift
public func <*> <C: CollectionType, T, U> (left: Parser<C, T -> U>.Function, right: Parser<C, T>.Function) -> Parser<C, U>.Function {
	return left >>- { $0 <^> right }
}
```

Seems to me though that what you've come up with feels a bit more natural given that functions aren't curried by default in Swift.

Any thoughts?

File: README.md [hunk 1]
```diff
@@ -104,16 +104,16 @@ See `Madness.playground` for some examples of parsing with Madness.
...
 
 	```swift
-	x --> { $0 }
+	x |> map { $0 }
+	{ $0 } <^> x
+	x --> { _, _, y in y }
 	```
 
-	parses `x` and maps its parse trees using the passed function. Use mapping to build your model objects.
+	parses `x` and maps its parse trees using the passed function. Use mapping to build your model objects. `-->` passes the input and parsed range as well as the parsed data for e.g. error reporting or AST construction.
```
Comment (sharplet) [inline README.md]: Thanks for updating this! :smile:

**Comment by robrix** [comment]
>    ```swift
public func <*> <C: CollectionType, T, U> (left: Parser<C, T -> U>.Function, right: Parser<C, T>.Function) -> Parser<C, U>.Function
```

Whoa, trippy—that’s actually the proper type of [`Applicative`’s `<*>`](http://hackage.haskell.org/package/base-4.8.0.0/docs/Control-Applicative.html#v:-60--42--62-)!

I gotta experiment with that a bit. I’d much rather use `Applicative`’s type if possible, but I worry that lifting parsers into the `Applicative` context using `pure` may be unidiomatic. Or maybe the parser types should already be lifted into the `Applicative` context…?

**Comment by sharplet** [comment]
Some food for thought—this is a really great example of constructing non-trivial parsers in an application style: http://www.serpentine.com/blog/2008/02/06/the-basics-of-applicative-functors-put-to-practical-work/.

This is giving me lots of ideas! 😊

**Comment by sharplet** [comment]
Perhaps this could help make things more idiomatic:

```swift
public func lift<C: CollectionType, T, U, V>(f: (T, U) -> V) -> Parser<C, T -> U -> V>.Function {
	return pure(curry(f))
}

public func lift<C: CollectionType, T, U, V, W>(f: (T, U, V) -> W) -> Parser<C, T -> U -> V -> W>.Function {
	return pure(curry(f))
}
```

Used like so:

```swift
let hex2 = lift(+) <*> hex <*> hex
```

:rocket:

**Comment by sharplet** [comment]
Ok, so just playing around with Lambda Calculus.playground, ended up with something cool:

```swift
enum Lambda: Printable {
  // ...

  static func variable(name: String) -> Lambda {
    return Variable(name)
  }

  static func abstraction(name: String)(_ lambda: Lambda) -> Lambda {
    return Abstraction(name, Box(lambda))
  }

  static func application(left: Lambda)(_ right: Lambda) -> Lambda {
    return Application(Box(left), Box(right))
  }

  // ...
}

typealias LamdaParser = Parser<String, Lambda>.Function

let lambda: LamdaParser = fix { term in
	let symbol                   = %("a"..."z")
	let variable:    LamdaParser = Lambda.variable    <^> symbol
	let abstraction: LamdaParser = Lambda.abstraction <^> (%"λ" *> symbol) <*> (%"." *> term)
	let application: LamdaParser = Lambda.application <^> (%"(" *> term)   <*> (%" " *> term <* %")")

	return variable <|> abstraction <|> application
}
```

**Comment by robrix** [comment]
Ahhh, `Applicative` at its best. This is pretty gorgeous, @sharplet.

**Comment by joshvera** [comment]
:+1: Good to merge @robrix?

**Comment by robrix** [comment]
@joshvera: Gonna poke around a little more first.

**Comment by robrix** [comment]
@sharplet: I’ve gone with the proper Applicative version of `<*>` and included your `lift` functions. Thank you!

**Comment by robrix** [comment]
@joshvera: I’m pretty content with this now but am curious to hear what you think about the use of Applicative/`pair` vs. the prior tupling behaviour of `<*>`.

**Comment by sharplet** [comment]
Great work @robrix, looks pretty cool!

**Comment by robrix** [comment]
Thanks @sharplet!

@joshvera: I’ll leave this to you to merge or otherwise as you see fit.

**Comment by joshvera** [comment]
Looks great!

**Comment by robrix** [comment]
:tada:

**Pull request merged**